core/lib/zksync_core/src/sync_layer/fetcher.rs

```diff
@@ -60,11 +104,75 @@ impl MainNodeFetcherCursor {
         };
 
         Ok(Self {
-            miniblock,
+            next_miniblock,
             l1_batch,
+            prev_miniblock_hash,
         })
     }
 
+    pub(super) fn advance(&mut self, block: FetchedBlock) -> Vec<SyncAction> {
+        assert_eq!(block.number, self.next_miniblock);
+
+        let mut new_actions = Vec::new();
+        if block.l1_batch_number != self.l1_batch {
+            assert_eq!(
+                block.l1_batch_number,
+                self.l1_batch.next(),
+                "Unexpected batch number in the next received miniblock"
+            );
```

Comment:
A DQ: do we trust the sender to always correctly send this block or anyone can connect to us via gossip, send random block l1 batch number and DoS the node?

Reply:
A very good question! The L1 batch number is covered by the block quorum certificate (proof of authenticity, jointly signed by the validators). So a node cannot fake this number w/o the quorum certificate becoming incorrect, which [will be detected](https://github.com/matter-labs/era-consensus/blob/9f58e688e70f277550b14cb2a9bc343c874bb0eb/node/actors/sync_blocks/src/peers/mod.rs#L466-L469) on the consensus side, and the block will be rejected. If the L1 batch number is incorrect with a correct quorum certificate, we have a compromised supermajority of validators, or there's something seriously wrong with the node (e.g., its deserialization logic is malfunctioning), so panicking seems like a natural step at this point.

Reply:
> we have a compromised supermajority of validators

So a compromised supermajority of validators will bring down all honest nodes which is theoretically exactly what they would want?

Reply:
If we have a compromised supermajority of validators, we have *far* more problems besides crashing non-validating nodes. E.g., they can issue different versions of the L2 blockchain to different nodes, stop processing some or all transactions altogether etc.

Reply:
My understanding of PoS (if we are aiming for one) is:

>  they can issue different versions of the L2 blockchain to different nodes

That is an equivocation and a proper PoS should have slashing mechanisms for that

> stop processing some or all transactions altogether etc.

That is a liveness check and dead validators should also be slashed.

Slashing would be impossible if all honest nodes are crashing, but I won't push this offtopic any further.